### PR TITLE
POC for Appointments API

### PR DESCRIPTION
**Pull request by dlloyd1-ch**
Simple RESTful API for retrieving an appointment from CHS mongoDB

Jira: SAPI-382

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,39 @@
...
+* A Mongodb instance running a database called appointments
+* A collection within it called appointments
+* Configure spring.data.mongodb.uri in application.properties to point to specific database
+
+## Start appointments API
+mvn package && java -jar target/poc-appointment-api-0.0.1-SNAPSHOT.jar
+
+or
+
+run PocAppointmentApiApplication.java from within an IDE
```
Comment (dlgroves) [inline README.md:24]: Or: mvn spring-boot:run

File: src/main/java/uk/gov/ch/pocappointmentapi/Appointment.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
+package uk.gov.ch.pocappointmentapi;
+
+import org.springframework.data.annotation.Id;
+import org.springframework.data.mongodb.core.mapping.Document;
+import org.springframework.data.mongodb.core.mapping.Field;
+
+import com.fasterxml.jackson.annotation.JsonIgnore;
+
+@Document(collection="appointments")
+public class Appointment {
```
Comment (dlgroves) [inline src/main/java/uk/gov/ch/pocappointmentapi/Appointment.java]: Could use Hibernate OGM with an EntityManager and JPA annotations to make this more portable.

File: src/main/java/uk/gov/ch/pocappointmentapi/Appointment.java [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+	}
+
+	@Override
+	public String toString() {
+		return "Appointment [appointmentId=" + appointmentId + ", officerId=" + officerId + ", companyNumber="
+				+ companyNumber + ", companyName=" + companyName + ", forename=" + data.forename + ", surname=" + data.surname
+				+ "]";
+	}
+	
+	class Data{
```
Comment (dlgroves) [inline src/main/java/uk/gov/ch/pocappointmentapi/Appointment.java]: Please move this inner instance class out into a separate file.

File: src/main/java/uk/gov/ch/pocappointmentapi/Appointment.java [hunk 4]
```diff
@@ -0,0 +1,226 @@
...
+		
+		public String getUpdatedAt() {
+			return updatedAt;
+		}
+
+		public void setUpdatedAt(String updatedAt) {
+			this.updatedAt = updatedAt;
+		}
+
+		class ServiceAddress{
```
Comment (dlgroves) [inline src/main/java/uk/gov/ch/pocappointmentapi/Appointment.java]: Please move this inner instance-inner instance class out into a separate file.

File: src/main/java/uk/gov/ch/pocappointmentapi/AppointmentController.java [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.ResponseStatus;
+import org.springframework.web.bind.annotation.RestController;
+
+@RestController
+public class AppointmentController {
```
Comment (dlgroves) [inline src/main/java/uk/gov/ch/pocappointmentapi/AppointmentController.java:13]: We could use JAX-RS annotations here for added portability - you may need additional dependencies and configuration for the Spring Boot plugin in pom.xml to support this, though.

File: src/main/java/uk/gov/ch/pocappointmentapi/AppointmentController.java [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+
+	@Autowired
+	AppointmentRepository repository;
+
+	@GetMapping(value = "/company/{companyNumber}/appointment/{appointmentId}")
+	public Appointment appointment(@PathVariable String companyNumber, @PathVariable String appointmentId) {
+		return repository.findByCompanyNumberAndAppointmentId(companyNumber, appointmentId);
```
Comment (dlgroves) [inline src/main/java/uk/gov/ch/pocappointmentapi/AppointmentController.java]: We would have to add another layer of abstraction here (e.g. Service object) to support transformation.

File: src/test/java/uk/gov/ch/pocappointmentapi/PocAppointmentApiTests.java [hunk 7]
```diff
@@ -0,0 +1,16 @@
+package uk.gov.ch.pocappointmentapi;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.boot.test.context.SpringBootTest;
+import org.springframework.test.context.junit4.SpringRunner;
+
+@RunWith(SpringRunner.class)
+@SpringBootTest
+public class PocAppointmentApiTests {
```
Comment (dlgroves) [inline src/test/java/uk/gov/ch/pocappointmentapi/PocAppointmentApiTests.java:10]: Our integration tests will eventually go here; please either move this to src/it/java and add src/it/java as an extra test source or put this in a separate submodule.

**Review by dlgroves — CHANGES_REQUESTED** [review]
Looking good, but please make the following changes before merging this:
* Please reformat all files to fix indentation issues and please use soft tabs (4 spaces).
* Please move inner instance classes out into separate class files.
* Please either move PocAppointmentApiTests into src/it/java and add this as a test source directory or to a submodule.

File: README.md (hunk 1 above)
Comment (dlloyd1-ch) [reply README.md:24]: Good shout.

File: src/main/java/uk/gov/ch/pocappointmentapi/Appointment.java (hunk 2 above)
Comment (dlloyd1-ch) [reply src/main/java/uk/gov/ch/pocappointmentapi/Appointment.java]: True, but given this just a POC, I don't think portability is a requirement.

File: src/main/java/uk/gov/ch/pocappointmentapi/Appointment.java (hunk 3 above)
Comment (dlloyd1-ch) [reply src/main/java/uk/gov/ch/pocappointmentapi/Appointment.java]: Done

File: src/main/java/uk/gov/ch/pocappointmentapi/Appointment.java (hunk 4 above)
Comment (dlloyd1-ch) [reply src/main/java/uk/gov/ch/pocappointmentapi/Appointment.java]: Done

File: src/main/java/uk/gov/ch/pocappointmentapi/AppointmentController.java (hunk 5 above)
Comment (dlloyd1-ch) [reply src/main/java/uk/gov/ch/pocappointmentapi/AppointmentController.java:13]: We could, but this is just a POC so trying to keep it as simple as possible.

File: src/main/java/uk/gov/ch/pocappointmentapi/AppointmentController.java (hunk 6 above)
Comment (dlloyd1-ch) [reply src/main/java/uk/gov/ch/pocappointmentapi/AppointmentController.java]: Yes, in the actual implementation we would need a service layer - but not needed for POC.

File: src/test/java/uk/gov/ch/pocappointmentapi/PocAppointmentApiTests.java (hunk 7 above)
Comment (dlloyd1-ch) [reply src/test/java/uk/gov/ch/pocappointmentapi/PocAppointmentApiTests.java:10]: This test file and it's location is generated by Spring-Boot. We would need extra entries in the pom to support the src/it/java source folder. I would say this is not needed for a POC.

**Review by dlgroves — APPROVED** [review]

**Pull request merged**